### PR TITLE
feat(history-service): support multiple location changes in one navigation

### PR DESCRIPTION
**Pull request by stemey**
- add RootHistoryService

### CODE REVIEW

File: packages/history-service/src/__tests__/create-root-location-transformer.test.ts [hunk 1]
```diff
@@ -242,4 +242,41 @@ describe('#createRootLocationTransformer', () => {
       });
     });
   });
+  describe('#createLocation', () => {
+    describe('without a primary', () => {
+      it('joins all consumer locations together as a single encoded query param ', () => {
```
Comment (clebert) [inline packages/history-service/src/__tests__/create-root-location-transformer.test.ts]: `joins all non-primary consumer locations together as a single encoded query param` and please remove the last space

File: packages/history-service/src/__tests__/create-root-location-transformer.test.ts [hunk 2]
```diff
@@ -242,4 +242,41 @@ describe('#createRootLocationTransformer', () => {
...
+
+        expect(rootLocation.pathname).toEqual('/');
+        expect(rootLocation.search).toEqual(
+          '---=%7B%22test%3A1%22%3A%22%2Ffoo%22%2C%22test%3A2%22%3A%22%2Fbar%22%7D'
+        );
+      });
+    });
+
+    describe('with a primary', () => {
+      it('joins all consumer locations together as a single encoded query param ', () => {
```
[the remark this answers is not in the document]
Comment (clebert) [reply packages/history-service/src/__tests__/create-root-location-transformer.test.ts]: and please remove the last space

File: packages/history-service/src/create-root-location-transformer.ts [hunk 3]
```diff
@@ -128,6 +137,29 @@ export function createRootLocationTransformer(
       return getConsumerPath(consumerPaths, consumerUid);
     },
```
Comment (clebert) [inline packages/history-service/src/create-root-location-transformer.ts]: I won't go further into stylistic problems, as this code has obviously not been TSLint checked and Prettier formatted. Please configure your IDE accordingly or use the available commands from package.json.

File: packages/history-service/src/create-root-location-transformer.ts [hunk 4]
```diff
@@ -128,6 +137,29 @@ export function createRootLocationTransformer(
...
 
+    createLocation(
+      ...consumerLocations: ConsumerLocation[]
+    ): history.Location {
+      if (options.primaryConsumerUid && !consumerLocations.some((consumerLocation) => consumerLocation.historyKey===options.primaryConsumerUid)) {
+        throw new Error('primary consumer is mandatory')
```
Comment (clebert) [inline packages/history-service/src/create-root-location-transformer.ts]: When writing the error messages, please follow the style that prevails in this project. The first letter is capitalized and we have punctuation.

File: packages/history-service/src/create-root-location-transformer.ts [hunk 5]
```diff
@@ -128,6 +137,29 @@ export function createRootLocationTransformer(
...
+        pathname:'/',
+        search:'',
+        state:{}
+      };
+      for (const consumerLocation of consumerLocations) {
+        rootlocation = this.createRootLocation(
+          consumerLocation.location,
+          rootlocation,
+          consumerLocation.historyKey
+        ) as history.Location;
```
Comment (clebert) [inline packages/history-service/src/create-root-location-transformer.ts]: This cast is not correct and obscures potential errors. Shouldn't the return value of the method rather be "LocationDescriptorObject"?

File: packages/history-service/src/create-root-location-transformer.ts (hunk 4 above)
Comment (clebert) [reply packages/history-service/src/create-root-location-transformer.ts]: And please add a test for this error, we expect your code to have 100% test coverage.

File: packages/history-service/src/create-root-location-transformer.ts [hunk 6]
```diff
@@ -22,6 +27,10 @@ export interface RootLocationTransformer {
     currentRootLocation: history.Location,
     consumerUid: string
   ): history.LocationDescriptorObject;
+
+  createLocation(
```
Comment (clebert) [inline packages/history-service/src/create-root-location-transformer.ts]: We should find a better name for this method. How about `createRootLocationForMultipleConsumers`?

File: packages/history-service/src/create-root-location-transformer.ts (hunk 5 above)
Comment (stemey) [reply packages/history-service/src/create-root-location-transformer.ts]: yes, we should change ConsumerLocation type. It should contain LocationDescriptorObject. We don't even support hash for any consumer but the primary. I am not sure if we should handle parameters and paths - path is usually enough and could be encoded much easier

File: packages/history-service/src/__tests__/create-root-location-transformer.test.ts [hunk 7]
```diff
@@ -242,4 +242,41 @@ describe('#createRootLocationTransformer', () => {
       });
     });
   });
+  describe('#createLocation', () => {
```
[the remark this answers is not in the document]
Comment (stemey) [reply packages/history-service/src/__tests__/create-root-location-transformer.test.ts]: done

File: packages/history-service/src/__tests__/create-root-location-transformer.test.ts [hunk 8]
```diff
@@ -242,4 +242,41 @@ describe('#createRootLocationTransformer', () => {
...
+    describe('without a primary', () => {
+      it('joins all consumer locations together as a single encoded query param ', () => {
+        const locationTransformer = createRootLocationTransformer({
+          consumerPathsQueryParamName: '---'
+        });
+
+        let rootLocation = locationTransformer.createLocation(
```
[the remark this answers is not in the document]
Comment (stemey) [reply packages/history-service/src/__tests__/create-root-location-transformer.test.ts]: all lint errors fixed

File: packages/history-service/src/__tests__/create-root-location-transformer.test.ts [hunk 9]
```diff
@@ -242,4 +242,41 @@ describe('#createRootLocationTransformer', () => {
...
+    });
+
+    describe('with a primary', () => {
+      it('joins all consumer locations together as a single encoded query param ', () => {
+        const locationTransformer = createRootLocationTransformer({
+          consumerPathsQueryParamName: '---',
+          primaryConsumerUid: 'test:1'
+        });
+
+        let rootLocation = locationTransformer.createLocation(
```
[the remark this answers is not in the document]
Comment (stemey) [reply packages/history-service/src/__tests__/create-root-location-transformer.test.ts]: all lint errors fixed

File: packages/history-service/src/__tests__/create-root-location-transformer.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (stemey) [reply packages/history-service/src/__tests__/create-root-location-transformer.test.ts]: done

File: packages/history-service/src/__tests__/create-root-location-transformer.test.ts (hunk 1 above)
Comment (stemey) [reply packages/history-service/src/__tests__/create-root-location-transformer.test.ts]: done

File: packages/history-service/src/create-root-location-transformer.ts (hunk 3 above)
Comment (stemey) [reply packages/history-service/src/create-root-location-transformer.ts]: done

File: packages/history-service/src/create-root-location-transformer.ts (hunk 5 above)
Comment (stemey) [reply packages/history-service/src/create-root-location-transformer.ts]: And what about the key in Location? We cannot properly multiplex that, I guess. It probably doesn't have a meaning for a consumer history. All appearances of Location in the consumerHistory should be cleaned up and a custom Location type without key should be used. I think the key is currently ignored completely.

File: packages/history-service/src/create-root-location-transformer.ts (hunk 6 above)
Comment (stemey) [reply packages/history-service/src/create-root-location-transformer.ts]: done

**Pull request closed without merging**